vm/src/obj/objbyteinner.rs

```diff
@@ -995,21 +995,89 @@ impl PyByteInner {
             .to_owned()
     }
 
-    pub fn split(&self, options: ByteInnerSplitOptions, reverse: bool) -> PyResult<Vec<&[u8]>> {
-        let (sep, maxsplit) = options.get_value()?;
-
-        if self.elements.is_empty() {
-            if !sep.is_empty() {
-                return Ok(vec![&[]]);
+    fn split_whitespace(&self, maxsplit: isize) -> Vec<Vec<u8>> {
+        let mut splited = Vec::new();
+        let mut count = maxsplit;
+        let mut haystack = &self.elements[..];
+        while let Some(offset) = haystack.find_byteset(Self::ASCII_WHITESPACES) {
+            if offset != 0 {
+                if count == 0 {
+                    break;
+                }
+                splited.push(haystack[..offset].to_vec());
+                count -= 1;
             }
-            return Ok(vec![]);
+            haystack = &haystack[offset + 1..];
+        }
+        if !haystack.is_empty() {
+            splited.push(haystack.to_vec());
         }
+        splited
+    }
 
-        if reverse {
-            Ok(split_slice_reverse(&self.elements, &sep, maxsplit))
+    pub fn split(
+        &self,
+        options: ByteInnerSplitOptions,
+        vm: &VirtualMachine,
+    ) -> PyResult<Vec<Vec<u8>>> {
+        let (sep, maxsplit) = options.get_value(vm)?;
+
+        let elements = if let Some(sep) = sep {
+            if maxsplit < 0 {
+                self.elements.split_str(&sep).map(<[u8]>::to_vec).collect()
+            } else {
+                self.elements
+                    .splitn_str(maxsplit as usize + 1, &sep)
+                    .map(<[u8]>::to_vec)
+                    .collect()
+            }
         } else {
-            Ok(split_slice(&self.elements, &sep, maxsplit))
+            self.split_whitespace(maxsplit)
+        };
+        Ok(elements)
+    }
+
+    fn rsplit_whitespace(&self, maxsplit: isize) -> Vec<Vec<u8>> {
+        let mut splited = Vec::new();
+        let mut count = maxsplit;
+        let mut haystack = &self.elements[..];
+        while let Some(offset) = haystack.rfind_byteset(Self::ASCII_WHITESPACES) {
+            if offset + 1 != haystack.len() {
+                if count == 0 {
+                    break;
+                }
+                splited.push(haystack[offset + 1..].to_vec());
+                count -= 1;
+            }
+            haystack = &haystack[..offset];
+        }
+        if !haystack.is_empty() {
+            splited.push(haystack.to_vec());
         }
+        splited
+    }
+
+    pub fn rsplit(
```

Comment:
Again maybe we can share code with `split`?

Reply:
I think this is possible to share but `split_whitespace` looks to require too much effort

Reply:
@palaviv how do you think about this new revision?

Reply:
Looks great